src/lib/envelope/message.ts

```diff
@@ -0,0 +1,67 @@
+import { v4 as uuid } from "uuid";
+
+// trying to not export this type, instead building types from this definition
+type MessagePayload = {
+    "text"  : {
+        displayName: string;
+        body: string;
+    };
+    "midi": {
+        state: 0 | 1;
+        number: number;
+        velocity: number;
+    };
+    "connect": {
+        userId: string;
+        userName: string;
+    };
+};
+
+type MessageType = keyof MessagePayload;
+
+export type Message<T extends MessageType = MessageType> = {
+    [K in T]: {
+        type: K;
+        payload: MessagePayload[K];
+        id: string;
+        userId: string;
+    }
+}[T];
+
+export type MidiMessage = Message<"midi">
+
+export type TextMessage = Message<"text">
+
+export type ConnectMessage = Message<"connect">
+
+export class MessageParser {
+    constructor(private readonly userId: string) { }
+
+    static decode<T extends MessageType>(s: string): Message<T> {
+        return JSON.parse(s);
+    }
+
+    decode<T extends MessageType>(s: string): Message<T> {
```

Comment:
yh can remove as not used atm

Reply:
I though it would ne good to create an instance in Jam.svelte but haven't gotten round to developing that idea further yet. I can bring it back if I think of something